capstone/src/main/java/com/google/sps/search/Trie.java

```diff
@@ -1,35 +1,122 @@
+   * name.
+   */
   public Trie() {
-    children = new HashMap<String, Trie>();
-    isEnd = false;
+    this.children = new HashMap<String, Trie>();
+    this.isName = false;
+    this.isFullName = false;
   }
 
+  /**
+   * Inserts name into trie, with full name being the last node appended to the end of the name.
+   *
+   * @param name user first name or last name
+   * @param fullName user full name.
+   */
   public void insert(String name, String fullName) {
     if (name.contentEquals("")) {
-      isEnd = true;
-      // add in the full name here
+      // indicates the end of first name or last name
+      this.isName = true;
+      // insert full name node as child
+      Trie fullNameNode = new Trie();
+      fullNameNode.isFullName = true;
+      this.children.put(fullName, fullNameNode);
     } else {
-      String firstChar = name.substring(0, 1);
-      if (!(children.containsKey(firstChar))) {
-        children.put(firstChar, new Trie());
+      String firstChar = name.substring(0, 1).toUpperCase();
+      if (!(this.children.containsKey(firstChar))) {
+        this.children.put(firstChar, new Trie());
       }
-      children.get(firstChar).insert(name.substring(1), fullName);
+      this.children.get(firstChar).insert(name.substring(1), fullName);
     }
   }
 
-  public Set<String> findAllWithPrefix() {
-    return null;
+  /**
+   * Returns set of strings with names that match the given prefix.
+   *
+   * @param prefix prefix to search in trie
+   * @param totalPrefix parameter needed to build string
+   * @return set of strings with matching prefix
+   */
+  public Set<String> searchWithPrefix(String prefix, String totalPrefix) {
+    if (prefix.equals("")) {
+      return findAll(new TreeSet<String>(), totalPrefix);
+    } else {
+      String firstChar = prefix.substring(0, 1).toUpperCase();
+      if (this.children.containsKey(firstChar)) {
+        return this.children.get(firstChar).searchWithPrefix(prefix.substring(1), totalPrefix);
+      } else {
+        return Collections.emptySet();
+      }
+    }
+  }
+
+  /**
+   * Helper method used to find all words from a certain node.
+   *
+   * @param names set of matching names
+   * @param prefix characters from the trie built up to current node
+   * @return set of full names formed from node in Trie
+   */
+  private Set<String> findAll(Set<String> names, String prefix) {
+    Map<String, Trie> possibilities = children;
+    /* If current node is the end of the a first name or last name, look through its map of children
+     * and add the full name nodes to set of names to be returned. */
+    if (this.isName) {
+      for (String child : possibilities.keySet()) {
+        StringBuilder sb = new StringBuilder();
+        if (possibilities.get(child).isFullName) {
+          sb.append(child);
+          names.add(sb.toString());
+        }
+      }
+    }
+    for (String letter : possibilities.keySet()) {
```

Comment:
With the `isFullName` approach, aren't you not only iterating over letters at this point but also full name nodes?

Reply:
Oh yeah I that bad naming then because the way I had it before the children could contain either single letter or full name. This solution still works for the way the code is written now, which is what you suggested -- having a set hold full names